crates/wasi-common/src/table.rs

```diff
@@ -10,83 +12,89 @@ use std::collections::HashMap;
 /// The `Table` type is intended to model how the Interface Types concept of Resources is shaping
 /// up. Right now it is just an approximation.
 pub struct Table {
-    map: HashMap<u32, Box<dyn Any + Send + Sync>>,
-    next_key: u32,
+    map: RwLock<HashMap<u32, Arc<dyn Any + Send + Sync>>>,
+    next_key: AtomicU32,
 }
 
 impl Table {
     /// Create an empty table. New insertions will begin at 3, above stdio.
     pub fn new() -> Self {
         Table {
-            map: HashMap::new(),
-            next_key: 3, // 0, 1 and 2 are reserved for stdio
+            map: RwLock::new(HashMap::new()),
+            next_key: 3.into(), // 0, 1 and 2 are reserved for stdio
         }
     }
 
     /// Insert a resource at a certain index.
-    pub fn insert_at(&mut self, key: u32, a: Box<dyn Any + Send + Sync>) {
-        self.map.insert(key, a);
+    pub fn insert_at<T: Any + Send + Sync>(&self, key: u32, a: Arc<T>) {
+        self.map.write().unwrap().insert(key, a);
     }
 
     /// Insert a resource at the next available index.
-    pub fn push(&mut self, a: Box<dyn Any + Send + Sync>) -> Result<u32, Error> {
+    pub fn push<T: Any + Send + Sync>(&self, a: Arc<T>) -> Result<u32, Error> {
         // NOTE: The performance of this new key calculation could be very bad once keys wrap
         // around.
-        if self.map.len() == u32::MAX as usize {
+        if self.map.read().unwrap().len() == u32::MAX as usize {
             return Err(Error::trap(anyhow::Error::msg("table has no free keys")));
         }
         loop {
-            let key = self.next_key;
-            self.next_key = self.next_key.wrapping_add(1);
-            if self.map.contains_key(&key) {
+            let key = self
+                .next_key
+                .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+            if self.map.read().unwrap().contains_key(&key) {
                 continue;
             }
-            self.map.insert(key, a);
+            self.map.write().unwrap().insert(key, a);
             return Ok(key);
         }
     }
```

Comment:
This should be all one atomic operation which can be protected by one lock, there's no need to use both atomics and locks here.